core/src/main/java/org/infinispan/container/impl/EntryFactoryImpl.java

```diff
@@ -228,22 +240,19 @@ private CacheEntry getFromContainerForRead(Object key, int segment, boolean isOw
       }
    }
 
-   private InternalCacheEntry innerGetFromContainer(Object key, int segment, boolean writeOperation, boolean returnExpired) {
-      InternalCacheEntry ice;
-      // Write operations should not cause expiration events to occur, because we will most likely overwrite the
-      // value anyways - also required for remove expired to not cause infinite loop
-      if (writeOperation) {
-         ice = container.peek(segment, key);
-         if (ice != null && !returnExpired && ice.canExpire()) {
-            long wallClockTime = timeService.wallClockTime();
-            if (ice.isExpired(wallClockTime)) {
-               ice = null;
-            } else {
-               ice.touch(wallClockTime);
+   private InternalCacheEntry innerGetFromContainer(Object key, int segment, boolean returnExpired, boolean isPrimaryOwner) {
+      InternalCacheEntry ice = container.peek(segment, key);
+      if (ice != null && !returnExpired) {
+         long currentTime = timeService.wallClockTime();
+         if (ice.isExpired(currentTime)) {
+            // This means it is a write operation that isn't expiration and we are the owner, thus we should
+            // actually expire the entry from memory
+            if (isPrimaryOwner) {
+               // This method is always called from a write operation
+               expirationManager.entryExpiredInMemory(ice, currentTime, true).join();
```

Comment:
I'm not sure about this, it feels like we're spreading the expiration logic in too many places. And wrapping entries shouldn't block...

Reply:
Where would you recommend the logic should go? I don't have a good spot since we don't want this in the `DataContainer`. Unless you think we should move the other logic out of the `DataContainer` into here?

Wrapping entries already blocks for max idle :)

Reply:
Why not in the `DataContainer`? It already deals with expiration...

Although moving the expiration logic somewhere where it could be made non-blocking does sound better.

Reply:
I responded quickly to run out, but the more I am thinking I would rather probably have this in `EntryFactoryImpl` and make both methods non blocking with interceptor stack. Although I haven't looked at details if this is possible.

Reply:
Now that I got a chance to look at this, I am not sure how to make this non blocking without changing the `EntryFactory` interface. What do you think @danberindei ? You have had a lot more experience changing these things.

Reply:
Yeah, making it non-blocking is a lot more involved. It's in an impl package so changing the interface is not a big deal, but I wouldn't push for it in 9.4.

Not sure if moving everything expiration-related to `EntryFactoryImpl` right now is feasible either, that's why my first thought was to keep everything in the data container.

Reply:
Yeah to add it to the `DataContainer` would require changing that interface, which is a lot more difficult to do. So I would rather move it into `EntryFactoryImpl` at some point and eventually try to push that to the `EntryWrappingInterceptor`, but I don't see a nice easy way to do that. I would say what is here is probably what will have to be for 9.4.0.

Reply:
Ok for now :)

Reply:
I agree.